homeassistant/components/cloud/backup.py

```diff
@@ -0,0 +1,180 @@
+"""Backup platform for the cloud integration."""
+
+from __future__ import annotations
+
+import base64
+import hashlib
+from pathlib import Path
+from typing import Any
+
+from aiohttp import ClientResponseError
+from hass_nabucasa import Cloud
+from hass_nabucasa.cloud_api import (
+    async_files_delete_file,
+    async_files_download_details,
+    async_files_list,
+    async_files_upload_details,
+)
+
+from homeassistant.components.backup import (
+    AgentBackup,
+    BackupAgent,
+    BackupAgentError,
+    read_backup,
+)
+from homeassistant.core import HomeAssistant, callback
+
+from .client import CloudClient
+from .const import DATA_CLOUD, DOMAIN
+
+_STORAGE_BACKUP = "backup"
+
+
+def b64md5(path: Path) -> str:
+    """Calculate the MD5 hash of a file."""
+    with open(path, "rb") as f:
+        file_hash = hashlib.md5()
+        while chunk := f.read(8192):
+            file_hash.update(chunk)
+    return base64.b64encode(file_hash.digest()).decode()
+
+
+async def async_get_backup_agents(
+    hass: HomeAssistant,
+    **kwargs: Any,
+) -> list[BackupAgent]:
+    """Return the cloud backup agent."""
+    return [CloudBackupAgent(hass=hass, cloud=hass.data[DATA_CLOUD])]
+
+
+class CloudBackupAgent(BackupAgent):
+    """Cloud backup agent."""
+
+    name = DOMAIN
+
+    def __init__(self, hass: HomeAssistant, cloud: Cloud[CloudClient]) -> None:
+        """Initialize the cloud backup sync agent."""
+        super().__init__()
+        self._cloud = cloud
+        self._hass = hass
+
+    @callback
+    def _get_backup_filename(self) -> str:
+        """Return the backup filename."""
+        return f"{self._cloud.client.prefs.instance_id}.tar"
+
+    async def async_download_backup(
+        self,
+        backup_id: str,
+        *,
+        path: Path,
+        **kwargs: Any,
+    ) -> None:
+        """Download a backup file.
+
+        :param backup_id: The ID of the backup that was returned in async_list_backups.
+        :param path: The full file path to download the backup to.
+        """
+        if not self._cloud.is_logged_in:
+            raise BackupAgentError("Not logged in to cloud")
+
+        if not await self.async_get_backup(backup_id):
+            raise BackupAgentError("Backup not found")
+
+        details = await async_files_download_details(
+            self._cloud,
+            storage_type=_STORAGE_BACKUP,
+            filename=self._get_backup_filename(),
+        )
+
+        resp = await self._cloud.websession.get(
+            details["url"],
+            raise_for_status=True,
+        )
+
+        file = await self._hass.async_add_executor_job(path.open, "wb")
+        async for chunk in resp.content.iter_any():
+            await self._hass.async_add_executor_job(file.write, chunk)
+
+        metadata = await self._hass.async_add_executor_job(read_backup, path)
+        if metadata.backup_id != backup_id:
+            raise BackupAgentError("Backup not found")
```

Comment:
I don't think this is OK in production code. `CloudBackupAgent.async_upload_backup` uploads the metadata, can't we download it here instead of downloading the entire backup, then validating it?

Reply:
This seems redundant.
This is checked on L81

Reply:
Removed, if this is needed, the caller should do it.